src/Workspaces/Remote/ServiceHub/Host/RemoteExportProvider.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.IO;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.CodeAnalysis.ExternalAccess.AspNetCore.Internal.EmbeddedLanguages;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.Composition;
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+internal class RemoteExportProvider
+{
+    internal static readonly ImmutableArray<Assembly> RemoteHostAssemblies =
+        MefHostServices.DefaultAssemblies
+            .Add(typeof(AspNetCoreEmbeddedLanguageClassifier).Assembly)
+            .Add(typeof(BrokeredServiceBase).Assembly)
+            .Add(typeof(IRazorLanguageServerTarget).Assembly)
+            .Add(typeof(RemoteWorkspacesResources).Assembly)
+            .Add(typeof(IExtensionWorkspaceMessageHandler<,>).Assembly);
+
+    private static ExportProvider? s_instance;
+    internal static ExportProvider ExportProvider
+    {
+        get
+        {
+            Contract.ThrowIfNull(s_instance, "Default export provider not initialized. Call InitializeAsync first.");
+            return s_instance;
+        }
+    }
+
+    public static async Task InitializeAsync(string localSettingsDirectory, CancellationToken cancellationToken)
+    {
+        var resolver = new Resolver(SimpleAssemblyLoader.Instance);
+        var assemblyPaths = RemoteHostAssemblies.SelectAsArray(static a => a.Location);
+
+        var cacheDirectory = Path.Combine(localSettingsDirectory, "Roslyn", "RemoteHost", "Cache");
+        var catalogPrefix = "RoslynRemoteHost";
+
+        s_instance = await ExportProviderBuilder.CreateExportProviderAsync(assemblyPaths, resolver, cacheDirectory, catalogPrefix, performCleanup: true, loggerFactory: null, cancellationToken).ConfigureAwait(false);
```

Comment:
should work, but would be good to verify - in vscode extension version updates change the actual extension folder, so the cache can never be re-used across updates.  But in VS I assume that is not true.  We do use the assembly write time as part of the key so I think it should be fine, but not as confident

Reply:
I'm not clear on the concern. It looks like the export provider builder is based completely on last file write times (in addition to the full path). If either of those have changed, we should get a new cache. Is there a concern where those checks aren't sufficient?